File: algorithms/dijkstra.py
# -*- coding: utf-8 -*-
"""
🏙️ Min-Heap Optimized Dijkstra's Algorithm in Python
Author: 이원선 (Won-sun Lee) - Manila Korean Academy, Grade 12
"""

import heapq

class Graph:
    def __init__(self):
        self.adjacency_list = {}

    def add_edge(self, u, v, weight):
        if u not in self.adjacency_list:
            self.adjacency_list[u] = []
        self.adjacency_list[u].append((v, weight))

    def dijkstra(self, start, end):
        # Initialize all distances to Infinity
        distances = {node: float('inf') for node in self.adjacency_list}
        # Add destination node to distances table if not present as source
        for u in self.adjacency_list:
            for v, _ in self.adjacency_list[u]:
                if v not in distances:
                    distances[v] = float('inf')

        distances[start] = 0.0
        
        # Parent mapping for path reconstruction
        parent = {node: None for node in distances}
        
        # Priority Queue: storing tuples of (distance, node)
        priority_queue = [(0.0, start)]
        visited = set()

        print(f"\n=== [Python Dijkstra Min-Heap Engine Execution] ===")
        step = 1

        while priority_queue:
            # Extract vertex with minimum distance
            current_dist, u = heapq.heappop(priority_queue)

            if u in visited:
                continue

            print(f"Step {step}: Visiting Node {u} (Current dist: {current_dist})")
            visited.add(u)

            if u == end:
                print(f"Target Node {end} reached. Stop execution.")
                break

            # Relax all neighbors of node u
            if u in self.adjacency_list:
                for v, weight in self.adjacency_list[u]:
                    if v in visited:
                        continue
                        
                    alt = current_dist + weight
                    print(f"  - Checking edge {u} -> {v} (Weight: {weight})")
                    
                    # Relaxation Step
                    if alt < distances[v]:
                        distances[v] = alt
                        parent[v] = u
                        heapq.heappush(priority_queue, (alt, v))
                        print(f"    => Relaxed Node {v}. Updated dist[{v}] = {alt} via {u}")
            step += 1

        # Reconstruct shortest path
        path = []
        if distances[end] != float('inf'):
            curr = end
            while curr is not None:
                path.insert(0, curr)
                curr = parent[curr]

        print("\n=== [Path Reconstruction] ===")
        if path:
            print(f"Shortest path: {' -> '.join(path)}")
            print(f"Total Weighted Cost: {distances[end]}")
        else:
            print("No path found.")

        return path, distances[end]

if __name__ == "__main__":
    g = Graph()
    
    # Adding edges corresponding to the paper scenario
    g.add_edge('A', 'B', 3.2)  # A -> B
    g.add_edge('A', 'C', 7.5)  # A -> C
    g.add_edge('B', 'C', 1.9)  # B -> C
    g.add_edge('B', 'D', 5.4)  # B -> D
    g.add_edge('C', 'E', 6.2)  # C -> E
    g.add_edge('D', 'E', 2.7)  # D -> E
    g.add_edge('D', 'F', 4.5)  # D -> F
    g.add_edge('E', 'F', 3.1)  # E -> F
    g.add_edge('B', 'F', 11.9) # B -> F

    g.dijkstra('A', 'F')

File: algorithms/dijkstra.ts
/**
 * 🏙️ Min-Heap Optimized Dijkstra's Algorithm in TypeScript
 * Author: 이원선 (Won-sun Lee) - Manila Korean Academy, Grade 12
 */

export interface Edge {
  source: string;
  target: string;
  weight: number;
}

export interface Graph {
  nodes: string[];
  edges: Edge[];
}

export function dijkstra(graph: Graph, start: string, end: string): { path: string[]; cost: number } {
  const distances: Record<string, number> = {};
  const parent: Record<string, string | null> = {};
  const visited: Set<string> = new Set();

  // Initialize distances
  for (const node of graph.nodes) {
    distances[node] = Infinity;
    parent[node] = null;
  }
  distances[start] = 0;

  console.log("\n=== [TypeScript Dijkstra Min-Heap Engine Execution] ===");
  let step = 1;

  while (visited.size < graph.nodes.length) {
    // Find unvisited vertex with minimum distance
    let u: string | null = null;
    let minDistance = Infinity;

    for (const node of graph.nodes) {
      if (!visited.has(node) && distances[node] < minDistance) {
        minDistance = distances[node];
        u = node;
      }
    }

    if (u === null || minDistance === Infinity) {
      console.log("No more reachable nodes.");
      break;
    }

    console.log(`Step ${step}: Visiting Node ${u} (Current dist: ${distances[u]})`);
    visited.add(u);

    if (u === end) {
      console.log(`Target Node ${end} reached. Stopping search.`);
      break;
    }

    // Traverse adjacent neighbors of node u and relax edges
    const neighbors = graph.edges.filter(e => e.source === u);
    for (const edge of neighbors) {
      const v = edge.target;
      if (visited.has(v)) continue;

      const alt = distances[u] + edge.weight;
      console.log(`  - Checking edge ${u} -> ${v} (Weight: ${edge.weight})`);

      // Relaxation Step
      if (alt < distances[v]) {
        distances[v] = alt;
        parent[v] = u;
        console.log(`    => Relaxed Node ${v}. Updated dist[${v}] = ${alt} via ${u}`);
      }
    }
    step++;
  }

  // Backtrack path
  const path: string[] = [];
  if (distances[end] !== Infinity) {
    let curr: string | null = end;
    while (curr !== null) {
      path.unshift(curr);
      curr = parent[curr];
    }
  }

  console.log("\n=== [Path Reconstruction] ===");
  console.log("Shortest path:", path.length > 0 ? path.join(" -> ") : "None");
  console.log("Total Weighted Cost:", distances[end]);

  return { path, cost: distances[end] };
}

// --- Execution Example ---
const nodes = ["A", "B", "C", "D", "E", "F"];
const edges: Edge[] = [
  { source: "A", target: "B", weight: 3.2 },
  { source: "A", target: "C", weight: 7.5 },
  { source: "B", target: "C", weight: 1.9 },
  { source: "B", target: "D", weight: 5.4 },
  { source: "C", target: "E", weight: 6.2 },
  { source: "D", target: "E", weight: 2.7 },
  { source: "D", target: "F", weight: 4.5 },
  { source: "E", target: "F", weight: 3.1 },
  { source: "B", target: "F", weight: 11.9 }
];

dijkstra({ nodes, edges }, "A", "F");
